src/ray/object_manager/test/pull_manager_test.cc

```diff
@@ -1078,6 +1088,97 @@ TEST_F(PullManagerWithAdmissionControlTest, TestPrioritizeWorkerRequests) {
   AssertNoLeaks();
 }
 
+TEST_P(PullManagerTest, TestTimeOut) {
+  auto prio = BundlePriority::TASK_ARGS;
+  if (GetParam()) {
+    prio = BundlePriority::GET_REQUEST;
+  }
+  auto refs = CreateObjectRefs(1);
+  auto oids = ObjectRefsToIds(refs);
+  std::vector<rpc::ObjectReference> objects_to_locate;
+  auto req_id = pull_manager_.Pull(refs, prio, &objects_to_locate);
+  ASSERT_EQ(ObjectRefsToIds(objects_to_locate), oids);
+  ASSERT_TRUE(pull_manager_.HasPullsQueued());
+
+  std::unordered_set<NodeID> client_ids;
+  client_ids.insert(NodeID::FromRandom());
+  pull_manager_.OnLocationChange(oids[0], client_ids, "", NodeID::Nil(), false, 0);
+  ASSERT_TRUE(pull_manager_.IsObjectActive(oids[0]));
+  ASSERT_EQ(num_send_pull_request_calls_, 1);
+
+  // Object has no locations now.
+  fake_time_ += 10;
+  pull_manager_.OnLocationChange(oids[0], {}, "", NodeID::Nil(), false, 0);
+  ASSERT_FALSE(timed_out_objects_.count(oids[0]));
+  fake_time_ += 601;
+  pull_manager_.Tick();
+  ASSERT_TRUE(timed_out_objects_.count(oids[0]));
+  timed_out_objects_.clear();
+  RAY_UNUSED(pull_manager_.CancelPull(req_id));
+
+  req_id = pull_manager_.Pull(refs, prio, &objects_to_locate);
+  pull_manager_.OnLocationChange(oids[0], client_ids, "", NodeID::Nil(), false, 0);
+  ASSERT_TRUE(pull_manager_.IsObjectActive(oids[0]));
+  // Object has no locations now but it is pending execution.
+  fake_time_ += 10;
+  pull_manager_.OnLocationChange(oids[0], {}, "", NodeID::Nil(), true, 0);
+  ASSERT_FALSE(timed_out_objects_.count(oids[0]));
+  fake_time_ += 601;
+  pull_manager_.Tick();
+  ASSERT_FALSE(timed_out_objects_.count(oids[0]));
+  // Object has no locations now and is no longer pending execution.
+  pull_manager_.OnLocationChange(oids[0], {}, "", NodeID::Nil(), false, 0);
+  fake_time_ += 601;
+  pull_manager_.Tick();
+  ASSERT_TRUE(timed_out_objects_.count(oids[0]));
+  timed_out_objects_.clear();
+  RAY_UNUSED(pull_manager_.CancelPull(req_id));
+
+  AssertNoLeaks();
+}
+
+TEST_P(PullManagerTest, TestTimeOutAfterFailedPull) {
```

Comment:
I'm not sure how to trigger the Python-level error without adding some sort of chaos testing to the gRPC/pubsub layer, but here is a unit test for the pull manager that captures the bug (I think).

Reply:
Can you create an issue to write a python test after we merge https://github.com/ray-project/ray/pull/19970?

Reply:
I think (hopefully) that can give us good signals in the future from Ci.